R/slide.R

```diff
@@ -208,129 +206,198 @@ epi_slide <- function(
     }
     return(f_wrapper)
   }
+  epi_slide_one_group_partial <- function(.data_group, .group_key, ...) {
+    epi_slide_one_group(
+      .data_group, .group_key, ...,
+      f_factory = f_wrapper_factory,
+      before = window_args$before,
+      after = window_args$after,
+      ref_time_values = .ref_time_values,
+      all_rows = .all_rows,
+      new_col_name = .new_col_name,
+      used_data_masking = used_data_masking,
+      time_type = time_type,
+      complete_only = .complete_only
+    )
+  }
 
-  # Computation for one group, all time values
-  slide_one_grp <- function(.data_group,
-                            .group_key, # see `?group_modify`
-                            ..., # `...` to `epi_slide` forwarded here
-                            f_factory,
-                            starts,
-                            stops,
-                            ref_time_values,
-                            all_rows,
-                            new_col_name) {
-    # Figure out which reference time values appear in the data group in the
-    # first place (we need to do this because it could differ based on the
-    # group, hence the setup/checks for the reference time values based on all
-    # the data could still be off):
-    o <- ref_time_values %in% .data_group$time_value
-    starts <- starts[o]
-    stops <- stops[o]
-    kept_ref_time_values <- ref_time_values[o]
-
-    f <- f_factory(kept_ref_time_values)
-
-    # Compute the slide values
-    slide_values_list <- slider::hop_index(
-      .x = .data_group,
-      .i = .data_group$time_value,
-      .starts = starts,
-      .stops = stops,
-      .f = f,
-      .group_key, ...
+  # If .x is not grouped, then the trivial group is applied: https://dplyr.tidyverse.org/reference/group_map.html
+  # `...` from top of `epi_slide` are forwarded to `.f` here.
+
+  # If every group takes the length(available_ref_time_values) == 0 branch in
+  # epi_slide_one_group, then we end up in a situation where the result as no
+  # new columns at all. This is a very fragile solution, I might even just
+  # remove it and error instead.
+  result <- group_modify(.x, epi_slide_one_group_partial, ..., .keep = FALSE)
+  if (ncol(result) == ncol(.x)) {
+    cli_warn(
+      "epi_slide: no new columns were created. This can happen if every group has no available ref_time_values.
+      In this case we return your epi_df but with an all-NA column 'slide_value' or what you provided in
+      `.new_col_name`. If your computation returned data.frames you may not get the expected column names.",
+      class = "epiprocess__epi_slide_no_new_columns"
```

Comment:
question: what happens if user tries `epi_slide(.window_size = 7, ~ sum(.x$value), .new_col_name = "value")`?  Do we warn + overwrite their overwriting with NAs?

Reply:
hm, good point, yea this will probably overwrite their old values. i'll add a check for that.

Reply:
Hm, I'm thinking: the user intention here is clearly to overwrite their original values, so overwriting is fine. 

I do however think that we should validate `.new_col_name` a little more than we do now. I'm leaning towards "error if `any(.new_col_name %in% c("geo_value", "time_value"))`". This way, we make sure epi_slide returns an epi_df, but allow people to overwrite their own columns if they wish (since we don't modify their input object, we're not at risk of them losing their values completely here). While it would be nice to warn them if they overwrite, those warnings would get very annoying for users that intend that, so I don't think we should go down the road of warning.